### DISCUSSION TITLE
vCard plugin export

### DISCUSSION THREAD

**Discussion by shad-lp** [ckulka/baikal-docker#57] [Q&A]
Hello !

In a first time let me thank you for this wonderful piece of software, I'm using your image for some years now and it's really effective.
Recently, I was thinking that doing a regular export of my vcf cards on my Android smartphone was probably not the smart way to backup my addressbook.
I fall upon this : https://sabre.io/dav/vcf-export-plugin/

It seems like Baikal got a feature to export all my contacts in a single vcf file.
I tried to use the same URL to see if this plugin is implemented in your image, but as a result I stumbled upon an error page :

![baikal_export_contacts](https://user-images.githubusercontent.com/43928955/137615084-156441ab-4e51-4318-b1df-bf772f947244.PNG)

Any thoughts on this ? @ckulka

**Comment by shad-lp** [comment]
Okay forget it, I found the URL to use :

https://baikal.xxx.ovh/dav.php/addressbooks/lawrence/default?export

lawrence being my username, I had to login before.

Any idea on how a script could use credentials to login and automate the backup of my contacts on a regular basis ?

**Comment by ckulka** [comment]
Hi @shadowking001 ,

good question... I haven't done it yet myself, but this reminds me of the #13. If I had to guess, you can just use either Basic or Digest authentication, depending on what you selected under "WebDAV authentication type" when you began with Baikal (I'd imagine you can look those up in the Baikal configuration page too):

![image](https://user-images.githubusercontent.com/4717228/138177738-fa4d8ede-b4df-46ec-a4b0-04f6bf4364b8.png)

You can then use wget or cURL in a bash script to download the VCF file that's scheduled with systemd or cron (depending on what your OS uses):

```bash
# Example using wget
wget --output-document="./default.vcf" --auth-no-challenge --http-user=username --http-password=pass https://baikal.xxx.ovh/dav.php/addressbooks/lawrence/default?export

# Example using cURL (Basic auth)
curl -O default.vcf --user username:pass https://baikal.xxx.ovh/dav.php/addressbooks/lawrence/default?export

# Example using cURL (Digest auth)
curl -O default.vcf --digest --user username:pass https://baikal.xxx.ovh/dav.php/addressbooks/lawrence/default?export
```

The maintainers and/or community over at <https://github.com/sabre-io/Baikal> might also be able to help out.

**Comment by tanrak** [comment]
@shadowking001 you can use the wget command like this:
```
wget --output-document="./default.vcf" --auth-no-challenge --http-user=username --http-password=pass https://baikal.xxx.ovh/dav.php/addressbooks/lawrence/default?export
```
the same for calendar (for example default.ics)

**Reply by ckulka** [reply]
Thanks @tanrak , I completed the answer with your code snippets 👍

**Reply by BobWs** [reply]
> @shadowking001 you can use the wget command like this:
> 
> ```
> wget --output-document="./default.vcf" --auth-no-challenge --http-user=username --http-password=pass https://baikal.xxx.ovh/dav.php/addressbooks/lawrence/default?export
> ```
> 
> the same for calendar (for example default.ics)

Do you know what the script is for "gzip command compresses files"?
I would like to gzip the backup files, but I'm not that familiar with scripts, can someone please help me out?

**Reply by ckulka** [reply]
You can use `gzip` for this:

```bash
# Export vcf file when secured with Basic auth
wget --output-document="./default.vcf" --auth-no-challenge --http-user=<user> --http-password=<password> https://baikal.example.com/dav.php/addressbooks/<user>/default?export

# Export vcf file when secured with Digest auth
curl -o default.vcf --digest --user <user>:<password> https://baikal.example.com/dav.php/addressbooks/<user>/default?

# Compress as default.vcf.gz and delete default.vcf
gzip default.vcf

# Compress as default.vcf.gz and do not delete default.vcf
gzip -k default.vcf

# In case you need this too: rename with the current date in the filename
mv default.vcf.gz default-$(date '+%Y-%m-%d').vcf.gz
# Easier though to specify "default-$(date '+%Y-%m-%d').vcf" as the filename when you run wget or cURL
```

**Reply by BobWs** [reply]
Thanks for explaining the gzip worked! 
One more question do you now how to lets say `keep` the last `5 backups` in the script? How do I put that in the script? 
TIA

**Reply by ckulka** [reply]
Sure thing, here you go:

```bash
find /path/to/directory/default-*.vcf.gz -type f | sort | head -n -5 | xargs -I {} rm -- {}
```

* `find /path/to/directory/default-* -type f` finds all backup files in the directory. I'd specify the full path starting with a `/` so that the current working directory of the shell doesn't matter

* `sort` sorts the files in alphabetical order and because of the date format, it's from oldest (top) to newest (bottom)

* `head -n -5` skips the last 5 entries in the listing, i.e. returning all but the 5 newest files

* `xargs -I {} rm -- {}` deletes all the files returned by `head`

Technically, `find` without `sort` would be sufficient too because `find` returns them already sorted, but afair that's not "guaranteed" or documented by the method. It's very unlikely that this behaviour changes, but the `sort` in there future proofs you no matter what.

**Reply by BobWs** [reply]
Thanks for your help I will give it a try!

**Reply by BobWs** [reply]
The script is working, Thanks for your support!
